addon/chrome/content/preferences/preferences.js

```diff
@@ -52,13 +52,29 @@ var gPrefWindow = {
     window.addEventListener("input", this);
 
     window.document.querySelectorAll('input[type=number]').forEach(item => {
+      item.addEventListener("keypress", i => {
```

Comment:
I'm going to remove the POSTDATA input, so no need to control negative value
just replace all the code for keypress with
```
         if (i.key.length === 1 && !/\d/.test(i.key)) {
           i.preventDefault();
         }
``` 

Add global event listener for keypress and change on the window,
move all handlers to separate function handleInpuntNumberEvent (or something similar) call this function from handleEvent.
```diff
  handleEvent(aEvent) {
+   if (item.type === "number") {
+     this.handleInpuntNumberEvent(aEvent);
+   }
    const item = aEvent.target;
    switch (aEvent.type) {
```
You can use this function from Customize Styles window by using ```window.opener.gPrefWindow.handleInpuntNumberEvent```

change all handlers argument from i to e or event

Reply:
> I'm going to remove the POSTDATA input, so no need to control negative value
> just replace all the code for keypress with
> 
> ```
>          if (i.key.length === 1 && !/\d/.test(i.key)) {
>            i.preventDefault();
>          }
> ```

Explain this 'POSTDATA' thing further.

Reply:
its the input with min="-1" from session.xhtml

Reply:
> You can use this function from Customize Styles window by using `window.opener.gPrefWindow.handleInpuntNumberEvent`

Inputs Customize Styles are in custom elements should just copy codes to there.

Reply:
> if (i.key.length === 1 && !/\d/.test(i.key)) {
>            i.preventDefault();
>          }

```
        if (!(i.altKey || i.ctrlKey || i.metaKey) && i.key.length === 1 && !/\d/.test(i.key)) {
          i.preventDefault();
        }
```
try this

Reply:
Gonna check max length here. So caret no jump around.
```
!(i.altKey || i.ctrlKey || i.metaKey) &&
          (i.key.length === 1 && !/\d/.test(i.key)||
          (editor.textLength == i.target.maxLength &&
            editor.selection.anchorOffset == editor.selection.focusOffset))
```

Reply:
```
...
          (editor.textLength == i.target.maxLength &&
            editor.selection.anchorOffset == editor.selection.focusOffset))
```
this also block arrow keys, we need to enable arrow keys with our without modifiers.
- arrow-up,  arrow-down: should by default change the input value
- arrow-left,  arrow-right: should move caret to let the user select part of the input

Reply:
This won't.

I have 
```
if (e.charCode == 0) return;
```
At beginning.

Whole thing:
```
if (e.charCode == 0) return;
        if (!(e.altKey || e.ctrlKey || e.metaKey) &&
          (!/\d/.test(e.key) ||
          (editor.textLength == e.target.maxLength &&
            editor.selection.anchorOffset == editor.selection.focusOffset))) {
          e.stopPropagation();
          e.preventDefault();
        }
```